drivers/usb_c/tcpc/rt1715.c

```diff
@@ -0,0 +1,725 @@
+{
+	const struct rt1715_cfg *cfg = dev->config;
+
+	return tcpci_tcpm_get_rp_value(&cfg->bus, rp);
+}
+
+static int rt1715_tcpc_set_cc(const struct device *dev, enum tc_cc_pull pull)
+{
+	const struct rt1715_cfg *cfg = dev->config;
+	struct rt1715_data *data = dev->data;
+
+	if (!data->initialized) {
+		return -EIO;
+	}
+
+	data->cc_changed = true;
+
+	return tcpci_tcpm_set_cc(&cfg->bus, pull);
+}
+
+static void rt1715_tcpc_set_vconn_discharge_cb(const struct device *dev,
+					       tcpc_vconn_discharge_cb_t cb)
+{
+	struct rt1715_data *data = dev->data;
+
+	data->vconn_discharge_cb = cb;
+}
+
+static void rt1715_tcpc_set_vconn_cb(const struct device *dev, tcpc_vconn_control_cb_t vconn_cb)
+{
+	struct rt1715_data *data = dev->data;
+
+	data->vconn_cb = vconn_cb;
+}
+
+static int rt1715_tcpc_vconn_discharge(const struct device *dev, bool enable)
+{
+	const struct rt1715_cfg *cfg = dev->config;
+
+	if (cfg->vconn_disc_gpio.port == NULL) {
+		/* RT1715 does not have built-in VCONN discharge path */
+		LOG_ERR("VCONN discharge GPIO is not defined");
+		return -EIO;
+	}
+
+	return gpio_pin_set_dt(&cfg->vconn_disc_gpio, enable);
+}
+
+static int rt1715_tcpc_set_vconn(const struct device *dev, bool enable)
+{
+	const struct rt1715_cfg *cfg = dev->config;
+	struct rt1715_data *data = dev->data;
+	int ret;
+
+	if (!data->initialized) {
+		return -EIO;
+	}
+
+	if (enable == true && cfg->vconn_ctrl_gpio.port != NULL) {
+		/* Enable external VCONN power supply */
+		gpio_pin_set_dt(&cfg->vconn_ctrl_gpio, true);
+	}
+
+	data->cc_changed = true;
+	ret = tcpci_tcpm_set_vconn(&cfg->bus, enable);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (enable == false && cfg->vconn_ctrl_gpio.port != NULL) {
+		/* Disable external VCONN power supply */
+		gpio_pin_set_dt(&cfg->vconn_ctrl_gpio, false);
+	}
+
+	if (data->vconn_cb != NULL) {
+		ret = data->vconn_cb(dev, data->cc_polarity, enable);
+	}
+
+	return ret;
+}
+
+static int rt1715_tcpc_set_roles(const struct device *dev, enum tc_power_role power_role,
+				 enum tc_data_role data_role)
+{
+	const struct rt1715_cfg *cfg = dev->config;
+
+	return tcpci_tcpm_set_roles(&cfg->bus, PD_REV30, power_role, data_role);
+}
+
+static int rt1715_tcpc_get_rx_pending_msg(const struct device *dev, struct pd_msg *msg)
+{
+	struct rt1715_data *data = dev->data;
+
+	/* Rx message pending? */
+	if (!data->msg_pending) {
+		return -ENODATA;
+	}
+
+	/* Query status only? */
+	if (msg == NULL) {
+		return 0;
+	}
+
+	/* Dequeue Rx FIFO */
+	*msg = data->rx_msg;
+	data->msg_pending = false;
+
+	/* Indicate Rx message returned */
+	return 1;
+}
+
+static int rt1715_tcpc_rx_fifo_enqueue(const struct device *dev)
+{
+	const struct rt1715_cfg *const cfg = dev->config;
+	struct rt1715_data *data = dev->data;
+	struct i2c_msg buf[5];
+	uint8_t rxbcnt;
+	uint8_t rxftype;
+	uint16_t rxhead;
+	uint8_t rx_data_size;
+	struct pd_msg *msg = &data->rx_msg;
+	int ret = 0;
+
+	buf[0].buf = TCPC_REG_RX_BUFFER;
```

Comment:
This isn't correct.  You're writing the TCPC register number into a pointer. 
```suggestion
	uint8_t tcpc_reg = TCPC_REG_RX_BUFFER;
	buf[0].buf = *tcpc_reg;
```

Reply:
my fault. fixed